File: app/src/main/java/roboyard/pm/ia/AGameState.java
package roboyard.pm.ia;

import java.util.ArrayList;

/**
 *
 * @author Pierre Michel
 */
public abstract class AGameState {
  
  public AGameState(AGameState parentState, IGameMove previousMove){
    this.derivedStates = new ArrayList<>();
    this.parentState = parentState;
    this.previousMove = previousMove;
    if(parentState != null){
      this.depth = parentState.depth+1;
    }
  }
  
  protected final ArrayList<AGameState> derivedStates;
  protected final AGameState parentState;
  protected final IGameMove previousMove;
  protected int depth=1;
  
}

File: shared/src/commonMain/kotlin/roboyard/logic/core/AGameState.kt
package roboyard.logic.core

import java.util.ArrayList

/**
 * 
 * @author Pierre Michel
 */
abstract class AGameState(
    protected val parentState: AGameState?,
    protected val previousMove: IGameMove?
) {
    protected val derivedStates: ArrayList<AGameState?>
    protected var depth: Int = 1

    init {
        this.derivedStates = ArrayList<AGameState?>()
        if (parentState != null) {
            this.depth = parentState.depth + 1
        }
    }
}
